src/gui/qgsgeometrywidget.cpp

```diff
@@ -0,0 +1,238 @@
+/***************************************************************************
+   qgsgeometrywidget.cpp
+    --------------------------------------
+   Date                 : March 2015
+   Copyright            : (C) 2015 Nyall Dawson
+   Email                : nyall.dawson@gmail.com
+***************************************************************************
+*                                                                         *
+*   This program is free software; you can redistribute it and/or modify  *
+*   it under the terms of the GNU General Public License as published by  *
+*   the Free Software Foundation; either version 2 of the License, or     *
+*   (at your option) any later version.                                   *
+*                                                                         *
+***************************************************************************/
+
+#include "qgsgeometrywidget.h"
+#include "qgsapplication.h"
+#include "qgsgeometry.h"
+#include "qgsjsonutils.h"
+#include <QLineEdit>
+#include <QHBoxLayout>
+#include <QToolButton>
+#include <QMenu>
+#include <QAction>
+#include <QClipboard>
+
+QgsGeometryWidget::QgsGeometryWidget( QWidget *parent )
+  : QWidget( parent )
+{
+  QHBoxLayout *layout = new QHBoxLayout( this );
+  layout->setContentsMargins( 0, 0, 0, 0 );
+  layout->setSpacing( 0 );
+
+  mLineEdit = new QLineEdit( this );
+  mLineEdit->setReadOnly( true );
+  mLineEdit->setStyleSheet( QStringLiteral( "font-style: italic; color: grey;" ) );
+  mLineEdit->setSizePolicy( QSizePolicy::MinimumExpanding, QSizePolicy::Minimum );
+  int width = mLineEdit->minimumSizeHint().width();
+  mLineEdit->setMinimumWidth( width );
+
+  mButton = new QToolButton( this );
+  mButton->setFixedSize( 30, 26 );
+  mButton->setStyleSheet( QStringLiteral( "QToolButton{ background: none; border: 1px solid rgba(0, 0, 0, 0%);} QToolButton:focus { border: 1px solid palette(highlight); }" ) );
+  mButton->setIconSize( QSize( 24, 24 ) );
+  mButton->setSizePolicy( QSizePolicy::Minimum, QSizePolicy::Minimum );
+  mButton->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/mActionOffsetCurve.svg" ) ) );
+
+  layout->addWidget( mLineEdit );
+  layout->addWidget( mButton );
+
+  setFocusProxy( mLineEdit );
+
+  mMenu = new QMenu( this );
+
+  mCopyWktAction = new QAction( mMenu );
+  mCopyWktAction->setText( tr( "Copy as WKT" ) );
```

Comment:
Is EWKT supported in copy and paste?

In general, I see that CRS is handled in `updateLineEdit` but not elsewhere.  

I'm not sure about the implications of supporting CRS in this widget.

Reply:
Not yet, no -- since ewkt interpretation depends on the database instance,  that would require some public data provider method for resolving strings to referenced geometries first.

Reply:
Basically what happens with the widget is:

- if no changes are made, then original crs is maintained and the geometry is unchanged. The attribute form never gets notified that the field value is changed.
- if a user changes the geometry by pasting wkt/geojson, then the crs will be reset. The postgres provider then gets a crs less geometry to store. On my tests this means that field's crs is automatically used (by logic somewhere in postgis)

To me that result is ok, and definitely better then the current fragile approach of going via the text edit widget. But we could further improve by :

- adding a crs selector/setter option to this widget to allow users to manually set the crs for changed geometries
- adding the crs (and geometry type) information to the field definition somehow. Right now we have no way of knowing the postgres field's crs and type constraints outside of the provider. Maybe a freeform metadata variant map in QgsField would work for this, or maybe a more formal/strict API is needed ...